### ISSUE TITLE
False positive: Class has no '__dataclass_fields__' member (no-member)

### ISSUE THREAD

**Issue by JulienPalard** [pylint-dev/pylint#9183]
### Bug description

```python
The code:


from dataclasses import dataclass


@dataclass
class Youpi:
    yapi: int


print(Youpi.__dataclass_fields__)
```

### Command used

```shell
pylint repro.py
```

### Pylint output

```python
$ pylint repro.py 
************* Module repro
repro.py:1:0: C0114: Missing module docstring (missing-module-docstring)
repro.py:5:0: C0115: Missing class docstring (missing-class-docstring)
repro.py:9:6: E1101: Class 'Youpi' has no '__dataclass_fields__' member (no-member)
```

### Expected behavior

Youpi is a dataclass, so it has a `__dataclass_fields__` attribute, yet pylint tells it has not.

At runtime it works.

In the other hand, neither the PEP nor the doc tells about `__dataclass_fields__`, so it could be considered "private" / "not guaranteed to be kept in future versions" / "unsafe to use".

I personally don't think so, by [Hyrum's Law](https://www.hyrumslaw.com/) people already depend on it. Also it's mentionned in typeshed: https://github.com/python/typeshed/blob/40caa050ce66a8dd62a62c4b4631767c4092d998/stdlib/_typeshed/__init__.pyi#L314

### Pylint version

```shell
$ pylint --version
pylint 2.17.4
astroid 2.15.6
Python 3.11.6 (main, Oct  8 2023, 05:06:43) [GCC 13.2.0]
```

### OS / Environment

Debian Trixie
